prototypes/palloc/pool_alloc.c

```diff
@@ -0,0 +1,212 @@
+#include <stdlib.h>
+#include <stdio.h>
+#include <string.h>
+#include <stdint.h>
+#include <unistd.h>
+#include "queue.h"
+#include "pool_alloc.h"
+
+#define align_to_binary_pow(x, a) ((((x) >> (a))+1) << (a))
+
+#define DEBUG 1
+
+#if defined(DEBUG) && DEBUG > 0
+    #define DEBUG_PRINT(text, args ...) fprintf(stderr, "DEBUG: %s:%d:%s(): " text, \
+        __FILE__, __LINE__, __func__, ##args)
+#else
+    #define DEBUG_PRINT(text, args ...)
+#endif
+
+#define panic(text, args ...)                                                 \
+  __extension__({                                               \
+    fprintf(stderr, "FATAL ERROR: %s:%d:%s(): " text, __FILE__, __LINE__, __func__, ##args);     \
+    exit(EXIT_FAILURE);                                         \
+  })
+  
+int page_size;
+
+typedef struct pool_item {
+    uint64_t pi_guard_number; //0xcafebabecafebabe by default, normally isn't changed, so if it has another value, memory corruption has taken place
+    pool_slab_t *pi_slab; //pointer to slab containing this item
+    uint64_t pi_data[0];
+} pool_item_t;
+
+pool_slab_t *create_slab(size_t size, void (*constructor)(void *, size_t)) {
+    DEBUG_PRINT("Entering create_slab\n");
+    pool_slab_t *slab = malloc(page_size); //TODO: Change this during implementation into kernel
+    slab -> ph_nused = 0;
+    slab -> ph_ntotal = (page_size-sizeof(pool_slab_t)-7)/(sizeof(pool_item_t)+size+1); // sizeof(pool_slab_t)+n*(sizeof(pool_item_t)+size+1)+7 <= page_size, 7 is maximum number of padding bytes for a bitmap 
+    
+    slab -> ph_nfree = slab -> ph_ntotal;
+    slab -> ph_start = sizeof(pool_slab_t)+align_to_binary_pow(slab -> ph_ntotal, 3);
+    memset(slab -> ph_bitmap, 0, slab -> ph_ntotal);
+    for (int i = 0; i < slab -> ph_ntotal; i++) {
+        pool_item_t *curr_pi = (pool_item_t*) ((uint64_t) slab+(slab -> ph_start)+i*(size+sizeof(pool_item_t)));
```

Comment:
You perform this computation (slab address to pool_item address) many times - consider wrapping it in a convenience macro, it would increase readability and decrease the chance of mistakes.